File: src/mcp-server/tools/research.ts
/** リサーチレポート保存ツール */

import { eq, desc } from "drizzle-orm";
import { db } from "../../db/client.js";
import { researchReports } from "../../db/schema.js";

export async function writeResearchReport(
  reportType: string,
  title: string,
  content: string,
  symbolsAnalyzed: string[] | null = null,
  keyFindings: Record<string, unknown> | null = null,
  sessionId = "",
): Promise<Record<string, unknown>> {
  const [report] = await db
    .insert(researchReports)
    .values({
      reportType,
      title,
      content,
      symbolsAnalyzed: symbolsAnalyzed ?? [],
      keyFindings: keyFindings ?? {},
      sessionId,
    })
    .returning();

  return {
    id: report.id,
    report_type: report.reportType,
    title: report.title,
    created_at: report.createdAt?.toISOString(),
  };
}

export async function getRecentReports(
  reportType: string | null = null,
  limit = 10,
): Promise<Record<string, unknown>> {
  let query = db
    .select()
    .from(researchReports)
    .orderBy(desc(researchReports.createdAt))
    .limit(limit);

  if (reportType) {
    query = query.where(
      eq(researchReports.reportType, reportType),
    ) as typeof query;
  }

  const rows = await query;

  return {
    reports: rows.map((r) => ({
      id: r.id,
      report_type: r.reportType,
      title: r.title,
      symbols_analyzed: r.symbolsAnalyzed,
      key_findings: r.keyFindings,
      created_at: r.createdAt?.toISOString(),
      content_preview: r.content.slice(0, 300),
    })),
    count: rows.length,
  };
}

File: mcp-server/src/claude_trade/tools/research.py
"""リサーチレポート保存ツール"""

from sqlalchemy import select

from claude_trade.db import async_session
from claude_trade.models import ResearchReport


async def write_research_report(
    report_type: str,
    title: str,
    content: str,
    symbols_analyzed: list[str] | None = None,
    key_findings: dict | None = None,
    session_id: str = "",
) -> dict:
    """リサーチレポートを DB に保存する

    Args:
        report_type: "premarket", "intraday", "eod", "sector", "macro"
        title: レポートタイトル
        content: マークダウン形式のレポート本文
        symbols_analyzed: 分析した銘柄リスト
        key_findings: 主要な発見事項 (JSON)
        session_id: Claude セッション ID
    """
    report = ResearchReport(
        report_type=report_type,
        title=title,
        content=content,
        symbols_analyzed=symbols_analyzed or [],
        key_findings=key_findings or {},
        session_id=session_id,
    )

    async with async_session() as session:
        session.add(report)
        await session.commit()
        await session.refresh(report)

    return {
        "id": str(report.id),
        "report_type": report.report_type,
        "title": report.title,
        "created_at": report.created_at.isoformat(),
    }


async def get_recent_reports(
    report_type: str | None = None,
    limit: int = 10,
) -> dict:
    """最近のリサーチレポートを取得する

    Args:
        report_type: レポートタイプでフィルタ (省略で全タイプ)
        limit: 取得件数 (デフォルト 10)
    """
    async with async_session() as session:
        stmt = select(ResearchReport).order_by(ResearchReport.created_at.desc()).limit(limit)
        if report_type:
            stmt = stmt.where(ResearchReport.report_type == report_type)

        result = await session.execute(stmt)
        reports = result.scalars().all()

    return {
        "reports": [
            {
                "id": str(r.id),
                "report_type": r.report_type,
                "title": r.title,
                "symbols_analyzed": r.symbols_analyzed,
                "key_findings": r.key_findings,
                "created_at": r.created_at.isoformat(),
                "content_preview": r.content[:300],
            }
            for r in reports
        ],
        "count": len(reports),
    }
